---
title: "Bioinformatics Analysis and Visualisation of Medical Genomics Data - Week 1"
author: "Rodrigo Arcoverde Cerveira"
date: '`r format(Sys.Date(), "%Y-%m-%d")`'
output: 
 html_document:
    toc: true
    depth: 2
    toc_float: true
    number_sections: true
 github_document:
    preview_html: false
knit: (function(inputFile, encoding) {
          rmarkdown::render(inputFile,
                            encoding = encoding, 
                            output_file = paste0(Sys.Date(), "_",
                              xfun::sans_ext(inputFile), ".html"),
                                output_dir = "../results/lab_book/")})
---

```{r global.options, include=FALSE}
#setting up figures and chunks messages

knitr::opts_knit$set(echo = TRUE,
                     fig.width = 6, fig.height = 5,
                     fig.align = "center",
                     root.dir = rprojroot::find_rstudio_root_file(),
                     out.width = 768,
                     fig.pos = "H",
                     warning = FALSE, 
                     message = FALSE)
knitr::opts_chunk$set(warning = FALSE,
                      message = FALSE,
                      fig.width = 6, fig.height = 5,
                     fig.align = "center",
                     out.width = 768,
                     fig.pos = "H")

result.dir <- paste0("results/",Sys.Date(),"/")

## creates result.dir with date in if not existent
ifelse(isFALSE(dir.exists(paste0("../",result.dir))), dir.create(paste0("../",result.dir),recursive = TRUE),"Result directory for today exists already!")

result.dir <- here::here(result.dir)
data.dir <- here::here("data")

options(stringsAsFactors = FALSE) 
```

# Load libraries

This R code chunk is loading several libraries and defining a custom color palette.

```{r load_libraries}
#Load libraries
# if you do not have libraries, they are located in either CRAN or Bioconductor
# libraries should be able to be automatically retrieved using renv::restore() function.
library(ggpubr)
library(airway)
library(SummarizedExperiment)
library(gridExtra)
library(ggsci)
library(ggprism)
library(tidyverse)
```

# Task 1 - Literature

1. Read the research article of the hands-on working group you are assigned to (see file “Student Groups.pdf” in shared folder General course material).
2. Answer the following questions
a. What is the medically relevant insight from the article?

_The authors succesufully identified transcriptomic signatures that most likely play an important role in lupus disease establishment and disease exacerbation. These signature genes were specific to different immune cell types, reflecting the cell-type-specific mechanisms underlying this disease. Finally, the disease-activity signatures were associated with the therapeutic response to belimumab, a drug used to treat lupus, highlighting the potential for developing targeted therapies based on the distinct transcriptomic signatures of disease establishment and exacerbation._

b. Which genomics technology/ technologies were used?

_The majority of the analysis were based on Bulk RNA-seq from different FACS sorted cells. In addition, they have also used previously published scRNA-seq and GWAS datasets._

3. Further related research questions
a. List and explain at least three questions/ hypotheses you can think of that extend the analysis presented in the paper.

_Are the adaptive immune signatures, such as the Th1 gene signature being associated to mucocutaneous lupus activity, linked to specific T clonotypes? A similar question can be posed for the B cells that had associated with disease signatures_

_Recently, common viral infections have been linked to autoimmune diseases, such as Epstein-Barr Virus and multiple sclerosis. Previously, this and other viruses have been linked to trigger lupus as well. Coud you verify these in your dataset?_

_Knowing all the samples were from Asians, the applicability to other populations is a limitation of the study. Even though they correlate some of their findings with an European ancestor, it would be better to explore or integrate this with a more diverse dataset. This would allow to potentially identify both population specific patterns of lupus but also common signatures across distinct populations._

b. [Optional] Devise a computational analysis strategy for (some of) the listed questions under 3a.

_Although bulk RNA-seq is not optimal for B cell receptor and T cell receptor analysis, since you do not have in depth sequencing of those. It is possible to use [TRUST4](https://www.nature.com/articles/s41592-021-01142-2) to reconstruct clonal lineages of those cells. With this data, it is possible to explore HV-HJ gene usage, clonality diversity, clonal expansion, and identify potential clones linked to either disease establishment or exacerbation. Due to the large number of adaptive immune receptor repertoire (AIRR) datasets for autoimmune diseases, one could also combine and or confirm the results found in your cohort with others._ 

_Similar to the suggestion above, although bulk RNA-seq is not optimal for viral identification, it is possible to explore that. [VIRTUS](https://doi.org/10.1093/bioinformatics/btaa859) is one of the options to identify viral genomes in human bulk RNA-seq. This way, it is possible to identify in which cell clusters a specific virus is more likely to be found and associated to a gene signature_ 

# Task 2-3

_Task two relies on creating a public github repo to add assignments. This is done and is publicly available as a deployed [website here](https://rodrigarc.github.io/riken_ki_assignments) or the [source code data here](https://github.com/rodrigarc/riken_ki_assignments)._

# Task 4 - Using R example datasets

1. Use the R internal CO2 dataset (“data(CO2)”).

```{r}
data(CO2)
```

2. Describe briefly the content of the CO2 dataset using the help function.

```{r}
help(CO2)
str(CO2)
```

_The CO2 dataset contains 84 rows and 5 columns, representing data from an experiment on frost resistance in Echinochloa crus-galli grass. It includes variables like 'Plant,' 'Type,' 'Treatment,' 'conc,' and 'uptake,' along with modeling formulas, axis labels, and measurement units._

3. What is the average and median CO2 uptake of the plants from Quebec and Mississippi?

```{r}
# calculate mean of CO2 uptake in Quebec and Mississipi
CO2 %>%
  group_by(Type) %>%
  summarise(co2_mean_uptake = mean(uptake))
# calculate median of CO2 uptake in Quebec and Mississipi
CO2 %>%
  group_by(Type) %>%
  summarise(co2_mean_uptake = median(uptake))
```

4. [Optional] In the “airway” example data from Bioconductor, how many genes are expressed in each sample? How many genes are not expressed in any sample?

```{r}
# Load the airway dataset
data("airway")

# Calculate the number of expressed genes/transcritps in each sample
expressed_genes_per_sample <- colSums(assay(airway) > 0)

# Calculate the number of genes not expressed in any sample
genes_not_expressed_in_any_sample <- sum(rowSums(assay(airway) > 0) == 0)

# Print the number of expressed genes for each sample
cat("Number of expressed genes for each sample:\n")
print(expressed_genes_per_sample)

cat("\nNumber of genes not expressed in any sample:", genes_not_expressed_in_any_sample, "\n")
```

_PS.: Important to notice that by the size of this dataset, those are probably transcripts rather than only genes._

# Task 5 - R functions

Task 5 - R Functions
1. Write a function that calculates the ratio of the mean and the median of a given vector.
This is a helpful measure to detect data with outlying values.
Note: See Reference for R language

```{r}
mean_to_median_ratio <- function(input_vector) {
  if (length(input_vector) == 0) {
    stop("Input vector is empty.")
  }
  
  mean_val <- mean(input_vector)
  median_val <- median(input_vector)
  
  if (median_val == 0) {
    warning("Median is zero. The ratio may not be defined.")
    return(NULL)
  }
  
  ratio <- mean_val / median_val
  return(ratio)
}

# Example usage:
input_vector <- c(1, 2, 3, 4, 5)
result <- mean_to_median_ratio(input_vector)
cat("Mean-to-Median Ratio:", result, "\n")

```

2. Write a function that ignores the lowest and the highest value from a given vector and calculate the mean.

```{r}
mean_extremes_removal <- function(input_vector) {
  if (length(input_vector) < 3) {
    stop("Input vector should have at least three elements.")
  }
  
  max_val <- max(input_vector)
  min_val <- min(input_vector)
  
  filtered_vector <- input_vector[input_vector != max_val & input_vector != min_val]
  
  mean_val <- mean(filtered_vector)
  return(mean_val)
}

# Example usage:
input_vector <- c(1, 2, 3, 4, 5)
result <- mean_extremes_removal(input_vector)
cat("Mean (without extremes):", result, "\n")

```

3. Read about piping from here:https://r4ds.had.co.nz/pipes.html#pipes (you don’t have
to learn everything, a basic understanding of the usage is enough). Write a short (max. 300 characters, no spaces) explanation of why, how, and when not to use pipes.


_Pipes (e.g., `%>%` in tidyverse or the base R `|>` pipe) are used to streamline and enhance code readability by chaining operations. However, avoid excessive nesting for complex logic, as it can reduce code clarity and debuggability. In addition, piping is streamline, it does not work if it involves many inputs and outputs._


4. Familiarize yourself with the apply-family of functions (apply, lapply, sapply etc.)
http://uc-r.github.io/apply_family/. Write a short explanation (max. 300 characters, no spaces) of why they could be useful in your work.

_The apply-family functions (e.g., apply, lapply, sapply) are useful for speeding up and improving the readability of repetitive data manipulation tasks, reducing the need for explicit loops._

# Task 6 - Basic visualization with R


1. Compare the distributions of the body heights of the two species from the 'magic_guys.csv' dataset graphically

```{r}
# load magic_guys dataset
magic_guys_df <- read.csv(paste0(data.dir,"/magic_guys.csv"))
```

a. using the basic 'hist' function as well as ‘ggplot’ and ‘geom_histogram’ functions from the ggplot2 package. Optimize the plots for example by trying several different 'breaks'. Note that ggplot2-based functions give you many more options for changing the visualization parameters, try some of them.

```{r}
# Identify species in the magic_guys dataset
unique(magic_guys_df$species)

# Separate data for two species (assuming you have a 'species' column) and plot using base hist() function
magic_guys_df %>%
  filter(species == "jedi") %>%
  pull(length) %>%
  hist()

magic_guys_df %>%
  filter(species == "sith") %>%
  pull(length) %>%
  hist()

# Create histograms using 'ggplot2' with 'geom_histogram'
ggplot(magic_guys_df, aes(x = length, fill = species)) +
  geom_histogram(alpha = 0.7, position = "identity") +
  geom_density(aes(y = ..count..), alpha = 0.7, color = "black") +
  scale_fill_aaas() +
  scale_color_aaas() +
  theme_prism(border = TRUE, base_fontface = "plain", base_line_size = 0.5) +
  theme(aspect.ratio = 1) +
  facet_wrap(~species, ncol = 2)
```
b. Do the same comparison as in a. but with boxplots. If you want to use the ggplot2-package, use the functions ‘ggplot’ and ‘geom_boxplot’.

```{r}
# plot a boxplot to compare the length/height of the two different species
ggplot(magic_guys_df, aes(y = length, x = species, fill = species)) +
  geom_boxplot() +
  scale_fill_aaas() +
  theme_prism(border = T, base_fontface = "plain", base_line_size = 0.5) +
  theme(aspect.ratio = 1) 
```
c. Save the plots with the ‘png’, ‘pdf’, and ‘svg’ formats. In which situation would you use which file format?

```{r }
# Save the plot in different formats
ggsave(path = result.dir, "hist_species.png", width = 6, height = 6, dpi = 300)  # PNG format for web
ggsave(path = result.dir, "hist_species.pdf", width = 6, height = 6)             # PDF format for high-quality printing
ggsave(path = result.dir, "hist_species.svg", width = 6, height = 6)
```


_Use `png` when you need a raster image format suitable for web or digital display. It's useful for online articles, presentations, or sharing on websites._

_Use `pdf` when you need a high-quality vector format that retains crispness when printed. PDFs are excellent for reports, publications, and printing on physical media._

_Use `svg` when you need a vector format that can be easily edited and scaled without loss of quality. SVGs are suitable for vector graphics editing software or embedding in web pages. However, I usually save as pdf since Illustrator and affinity do not seem to like the format that R saves the svg_

2. Load the gene expression data matrix from the ‘microarray_data.tab’ dataset provided in the shared folder, it is a big tabular separated matrix.

```{r}
# load microarray_data dataset
microarray_df <- read.table(paste0(data.dir,"/microarray_data.tab"), header = TRUE, sep = "\t")
```
a. How big is the matrix in terms of rows and columns?

```{r}
# Get the number of rows and columns in the data frame
microarray_dim <- dim(microarray_df)
# Extract the number of rows and columns
num_rows <- microarray_dim[1]
num_columns <- microarray_dim[2]
# Print the number of rows and columns
cat("Number of Rows:", num_rows, "\n")
cat("Number of Columns:", num_columns, "\n")
```

b. Count the missing values per gene and visualize this result.
```{r}
# Count missing values (NA) per gene
missing_values_per_gene <- colSums(is.na(microarray_df))
# Create a data frame for visualization
missing_values_data <- data.frame(Gene = names(missing_values_per_gene),
                                  Missing_Values = missing_values_per_gene)
# Create a bar plot using ggplot2
missing_values_data %>%
  ggplot(aes(x = reorder(Gene, -Missing_Values), y = Missing_Values)) +
  geom_bar(stat = "identity", color = "lightblue") +
  labs(title = "Missing Values per Gene", x = "Gene label", y = "Count of Missing Values") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  scale_x_discrete(labels = function(x) ifelse(seq_along(x) %% 100 == 0, x, "")) +
  theme_prism(border = F, base_fontface = "plain", base_line_size = 0.5) +
  theme(axis.ticks.x = element_blank()) 
```

c. Find the genes for which there are more than X% (X=10%, 20%, 50%) missing values.

```{r}

# Set the desired percentage thresholds
thresholds <- c(0.1, 0.2, 0.5)  # Thresholds as proportions

# Calculate the total number of samples (columns) in the data frame
total_samples <- ncol(microarray_df)

# Create a list of data frames, one for each threshold
threshold_data_list <- lapply(thresholds, function(threshold) {
  # Count missing values (NA) per gene for the current threshold
  missing_values_per_gene <- colSums(is.na(microarray_df)) > threshold * total_samples
  # Create a data frame for the current threshold
  threshold_data <- data.frame(
    Gene = names(missing_values_per_gene),
    Missing_Values = missing_values_per_gene,
    Threshold = threshold
  )
  return(threshold_data)
})

# Combine the list of data frames into one data frame
gene_thresholds <- do.call(rbind, threshold_data_list) %>%
  group_by(Threshold, Missing_Values) %>%
  summarise(gene_count = n()) %>%
  ungroup()

gene_thresholds
```

_To avoid having large tables in this markdown, there is only printed the number of genes with missing values for each threshold. But if neeed, the genes could be explore on the `threshold_data_list`_

d. Replace the missing values by the average expression value for the particular
gene. (Note: Imputing data has to be used with caution!)

```{r}
# Replace missing values with the average expression value for each gene
microarray_df_imputed <- microarray_df %>%
  mutate(across(everything(), ~ifelse(is.na(.), mean(., na.rm = TRUE), .)))

# Verify that missing values have been replaced
microarray_df_imputed[1:5, 1:5]
```

_You can confirm that the previous `NA` values in the g2 column were replaced with the mean._ 

3. Visualize the data in the CO2 dataset in a way that gives you a deeper understanding
of the data. What do you see?

```{r}
# plot CO2 dataset with grouping based on treatment and wrapped with type
ggplot(CO2, aes(x = conc, y = uptake, color = Treatment, group = Treatment)) +
  geom_point() +
  geom_smooth() +
  labs(title = "CO2 Concentration vs. Uptake", x = "CO2 Concentration", y = "Uptake") +
  facet_wrap(~Type) +
  scale_color_manual(values = c("#EE0000B2", "#3B4992B2")) +
  theme_prism(border = T, base_fontface = "plain", base_line_size = 0.5) +
  theme(aspect.ratio = 1) 

pal_aaas( alpha = 0.7)(2)
```

_Plants from Quebec and Mississippi exhibit distinctive carbon dioxide (CO2) uptake profiles regardless of temperature conditions. Quebec plants consistently demonstrate higher CO2 uptake rates than their Mississippi counterparts. Temperature sensitivity reveals Mississippi plants to be more influenced by chilling, experiencing a drastic reduction in uptake, primarily beyond 250 ml/L CO2 concentration. In contrast, Quebec plants show greater resilience to temperature variations. This divergence may be attributed to their differing climates, with colder temperatures in Quebec potentially driving natural selection for plants with high CO2 uptake capabilities under cold conditions. These findings could emphasize the remarkable adaptability and regional distinctions in plant CO2 assimilation strategies. However, more information about the dataset, plant species, and reserch questions prior the experiment should be taken into account before drawing further conclusions._

# Task 7 - Genomics data

1. Install the Tidybiology package, which includes the data ‘chromosome’ and ‘proteins’
       devtools::install_github("hirscheylab/tidybiology")
       
```{r, eval=FALSE}
# Install tidybiology if not yet installed
devtools::install_github("hirscheylab/tidybiology")
```

```{r}
# load library and needed datasets
library(tidybiology)
data("chromosome")
data("proteins")
```

a. Extract summary statistics (mean, median and maximum) for the following variables from the ‘chromosome’ data: variations, protein coding genes, and miRNAs. Utilize the tidyverse functions to make this as simply as possible.

```{r}
# Print colnames in the chromosome dataset
colnames(chromosome)
# Filter the selected columns and calculate the mean using the summary function
chromosome %>% 
  select(variations, protein_codinggenes, mi_rna) %>% 
  summary()
```

b. How does the chromosome size distribute? Plot a graph that helps to visualize this by using ggplot2 package functions.

```{r}
# Plot the length in mm of each chromosome
chromosome %>%
  ggplot(aes(x = id, y = length_mm)) +
    geom_bar(stat = "identity", fill = "#3B4992B2", color = "black") +
    labs(title = "Chromosome Size Distribution", x = "Chromosome Number", y = "Length (mm)") +
    theme_prism(border = T, base_fontface = "plain", base_line_size = 0.5) +
    theme(aspect.ratio = 1,
          axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) 
```

c. Does the number of protein coding genes or miRNAs correlate with the length of the chromosome? Make two separate plots to visualize these relationships.

```{r}
# Plot a scatter and linear regression model for the protein coding gene counts and chromosome size
g1 <- chromosome %>%
  ggplot(aes(x = protein_codinggenes, y = length_mm)) +
    geom_point() +
    geom_smooth(method = "lm", color = "#EE0000B2") +
    stat_cor(method = "pearson") +
    labs(x = "Protein Coding Genes", y = "Chromosome Size") +
    theme_prism(border = T, base_fontface = "plain", base_line_size = 0.5) +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) 
# Plot a scatter and linear regression model for miRNA counts and chromosome size
g2 <- chromosome %>%
  ggplot(aes(x = mi_rna, y = length_mm)) +
    geom_point() +
    geom_smooth(method = "lm", color = "#EE0000B2") +
    stat_cor(method = "pearson") +
    labs(x = "miRNAs", y = "Chromosome Size") +
    theme_prism(border = T, base_fontface = "plain", base_line_size = 0.5) +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) 

# Arrange both plots side by side
grid.arrange(g1, g2, ncol = 2)
```

d. Calculate the same summary statistics for the ‘proteins’ data variables length and mass. Create a meaningful visualization of the relationship between these two variables by utilizing the ggplot2 package functions. Play with the colors, theme- and other visualization parameters to create a plot that pleases you.

```{r}
# Inspeact proteins dataset
str(proteins)

# Print summary of those variables
proteins %>% select(length, mass) %>% summary()

# Plot the proteins
proteins %>%
  ggplot(aes(x = length, y = mass)) +
  geom_bin2d() +
    scale_x_log10() +
    scale_y_log10() +
    geom_smooth(method = "lm", color = "#EE0000B2") +
    stat_cor(method = "pearson") +
    scale_fill_viridis_c(option = "inferno") +
    labs(title = "Relationship between Length and Mass in Proteins", x = "Length", y = "Mass") +
    theme_prism(border = T, base_fontface = "plain", base_line_size = 0.5) +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
          aspect.ratio = 1) 
```

# Session info

```{r session_info}
sessionInfo()
```
